### PR TITLE
Parameter Optimization

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
This program finds good parameters for an HTM system.

I wrote this program a while ago, and now I'd like to contribute it.

This is a large addition, and it is not ready yet.  I do **not** recommend reviewing the code, instead just read the documentation at the top of the file.

### CODE REVIEW

**Comment by breznak** [comment]
> This program finds good parameters for an HTM system.

hey, this is a huge and much welcome contribution! :100: 
It's neeeded for everyday practical usecases, #391 , and Hotgym, where our TM has insufficient representation (I don't know if there's an issue for that). 

Sorry, I'll get later to review, even just to read the doc. 

Some high-level thoughts about parameter optimization/tuning: 
- I don't know if this should be in this repo, or rather stand-alone nupic.swarming ? 
  - that way, other projects can also benefit
  - we'd still ensure OUR API is supported. 
- can we properly transform/describe our model parameters and leave the optimization to 3rd party tools (scikitlearn,...)? 
- see #205 for my comment about OPF support, that does the same thing. 
  - I guess your code has cleaner implementation
  - but OPF has a popular (?) API which we said to support? 
- if the code is working, I'd like to merge it ASAP, with potential future rework/obsoletion.

File: py/src/nupic/optimization/ae.py [hunk 1]
```diff
@@ -0,0 +1,898 @@
...
+calculates the confidence that a parameter change caused the score to change.
+ * It is extensible: new methods for automated parameter optimization can be
+added.  Currently this implements a basic grid search strategy.  In the future I
+hope to add a particle swarming method.
+
+To use this module, structure experiments as follows:
+    ExperimentModule is a python3 module containing the model to be optimized as
+    well as code to evaluate model performance.
+
+    ExperimentModule.default_parameters = {}
```
Comment (breznak) [inline py/src/nupic/optimization/ae.py]: as this is not compliant the OPF API, is there an industrial (really used) API for model parameter optimization? It'd be worth using that.

File: py/src/nupic/optimization/ae.py [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+    This global dictionary contains all of the parameters to modify.
+    Parameters must be one of the following types: dict, tuple, float, int.
+    Parameters can be nested in multiple levels of dictionaries and tuples.
+
+    ExperimentModule.main(parameters=default_parameters, argv=None, verbose=True)
+    Returns (float) performance of parameters, to be maximized.
+
+Usage: $ ae.py [ae-arguments] ExperimentModule.py [experiment-arguments]
```
Comment (breznak) [inline py/src/nupic/optimization/ae.py]: can you provide a runnable demo example? That would be worth 1000 words for first-timers wishing to use the code. 
MNIST would be an ideal candidate and you've already optimized that code for me once, so you should have the "experiment" files ready.

File: py/src/nupic/optimization/ae.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/ae.py]: There is not to my knowledge any standard interface for this.

This however is a nice and simple python API.

Comment (breznak) [reply py/src/nupic/optimization/ae.py]: https://www.khronos.org/nnef 
This is interesting, but seems like it's focused only at deep learning.

File: py/src/nupic/optimization/ae.py [hunk 3]
```diff
@@ -0,0 +1,898 @@
...
+    return min(lab.experiments, key=lambda x: x.attempts + rnd()).parameters
+
+
+def evaluate_best(lab):
+    rnd = random.random
+    best = max(lab.experiments, key = lambda X: X.mean() + rnd() )
+    return best.parameters
+
+
+class GridSearch(object):
```
Comment (breznak) [inline py/src/nupic/optimization/ae.py]: do you need to write the hyper param optimization code yourself? It is available in many toolkits? 
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

File: py/src/nupic/optimization/ae.py [hunk 4]
```diff
@@ -0,0 +1,898 @@
...
+import sys
+import shutil
+import random
+import pprint
+import time
+import datetime
+import tempfile
+import multiprocessing
+import resource
+import signal # Probably not X-platform ...
```
Comment (breznak) [inline py/src/nupic/optimization/ae.py]: would this be a problem? but hey..worst case, the feature will be avail only for certain OSs

File: py/src/nupic/optimization/swarming.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+import argparse
+import sys
+import os
+import random
+import pprint
+
+from .nupic.optimization.parameter_set import ParameterSet
+
+class ParticleSwarmOptimizations:
```
Comment (breznak) [inline py/src/nupic/optimization/swarming.py]: I do really like the recent changes! :clap: , extracting the particular search mechanism away, and adding the new PSO.

**Comment by dkeeney** [comment]
This is really cool.  🥇

File: py/src/nupic/optimization/swarming.py (hunk 5 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: > extracting the particular search mechanism away

This turned out to be more problematic than I expected.  Each search method needs:
1) Command line arguments
2) Queue experiments
3) Collect results, asynchronously from the queuing them

Currently it can only queue experiments.  I'm thinking about how to do all of this in a sane way that allows for new methods to plug-in.  I might make a new interface class to help with this.

File: py/src/nupic/optimization/ae.py (hunk 4 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/ae.py]: I looked this up, This use of `signal` can be replaced with multithreading.timer which _is_ X-platform.

File: py/src/nupic/optimization/swarming.py (hunk 5 above)
Comment (breznak) [reply py/src/nupic/optimization/swarming.py]: have you considered the 3rd party for param optimization, model selection? And "just" write enabling interface for our params & constructors?

Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: > https://www.khronos.org/nnef

I don't think this framework is useful, it's all about sharing trained deep learning networks.

> https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

Scikit learn appears very limited. 
- The only search method it has is GridSearch.  It does not do swarming.
- I also want the parameter optimization code to save all of the outputs of the program.  Which scikitlearn doesn't do.

Comment (breznak) [reply py/src/nupic/optimization/swarming.py]: agreed on NNEF 

> Scikit learn appears very limited. 

it was only a first hit I got from memory, these "python parameter optimization frameworks" look interesting: 
https://hyperopt.github.io/hyperopt/ 
https://optuna.org/ 
I do like your solution, but in the longrun I hope we'll have an interface to a 3rd party project, as there's lots of details which are out of scope for us (same as Serialization, ..)

Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: > have an interface to a 3rd party project, as there's lots of details which are out of scope for us

I understand, I will do some more research into existing solutions.

Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: > 3rd party solutions

I found a great project, it's called [nni](https://github.com/Microsoft/nni).  It is a framework for optimizing artificial neural networks (the non-biological kind).  I read through it and attempted (but ultimately failed) to run the MNIST example with it.
- Its open source (MIT license) so we could easily incorporate it into our project.
- It's run by microsoft, and they appear to be actively developing it. They accept external PR's and they might even fix issues we have with it.
- It is well designed.  It's got some neat features too.

All that praise said, I don't think it's ready for mainstream use.  It's rough at the edges.  The project is about a year old.  In a year this could be the defacto standard tool, but today I'd rather keep using my tools which I'm familiar with.

Comment (breznak) [reply py/src/nupic/optimization/swarming.py]: > I found a great project, it's called nni.  

I like NNI! although the whole AutoML seems overkill for us (we don't need network architecture "evolution", as we have HTM, we "just" need a way for param optimization. 

> I read through it and attempted (but ultimately failed) to run the MNIST example with it.

ok, let's go with what you have now. I would suggest not "wasting" time developing too many features, as long term we'd like to switch over to an external framework to do the work for us. 

The current will be a great improvement, so let's :+1:  

PS: what have failed running their mnist? Do you think it's not ready yet, or should I try in some later PR? 

Ideally, if you can take their (NNI or other suitable) design in mind, and develop the format for "experiment description" what you call it, similar/same, so we can later easily switch?

Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: > PS: what have failed running their mnist?

I tried running our MNIST in their framework.

> the format for "experiment description"

This is highly dependent on the specifics of the optimization framework.

Comment (ctrl-z-9000-times) [reply py/src/nupic/optimization/swarming.py]: I will push the work i did for NNI in a branch so that you can play with it.

I think the problem i had is that NNI does not start by trying the default parameters.  It spent a considerable amount of time trying parameters which obviously wont work.  Swarming has an advantage over TPE (the search method nni defaults to) in that it starts near the default parameter values which are at least sane.  The NNI framework has a way to end trials/experiments early if they obviously arent working, but i didnt use it.

**Comment by ctrl-z-9000-times** [comment]
I finished my review. The documentation still needs work before this can merge.
I added a lot of TODO notes for issues/bugs and feature requests.

File: py/src/nupic/examples/mnist.py [hunk 6]
```diff
@@ -87,16 +87,16 @@ def encode(self, image):
...
-    'dutyCyclePeriod': 1402,
-    'localAreaDensity': 0.024,
-    'minPctOverlapDutyCycle': 0.001,
-    'potentialPct': 0.2,
-    'stimulusThreshold': 6,
-    'synPermActiveInc': 0.01,
-    'synPermConnected': 0.422,
-    'synPermInactiveDec': 0.005
+    'boostStrength': 7.80643753517375,
+    'columnDimensions': (35415,),
```
Comment (breznak) [inline py/src/nupic/examples/mnist.py]: OT: I wonder how much the scores differ? It's said HTM is/should be more resistant to param changes. These seem like a rather big difference.

File: py/src/nupic/optimization/ae.py [hunk 7]
```diff
@@ -423,134 +427,144 @@ def save(self):
...
+                    X = self.get_experiment( trial.parameters )
+                    trial.collect_journal( X )
+                    trial.collect_score( X )
+                    # Notify the parameter optimization method that the
+                    # parameters which it suggested have finished evaluating.
+                    self.method.collect_results( X.parameters, trial.score )
+                    self.save()     # Write the updated Lab Report to file.
+
+
+class Worker(Process):
```
Comment (breznak) [inline py/src/nupic/optimization/ae.py]: nit, do you want to split some of these to separate classes?

File: py/src/nupic/examples/mnist.py (hunk 6 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/examples/mnist.py]: IIRC accuracy improved from about 90% to 96%.

**Comment by ctrl-z-9000-times** [comment]
Thanks for reviewing this!
